p2p/net/swarm/swarm_dial.go

```diff
@@ -398,21 +402,35 @@ func (s *Swarm) dialAddrs(ctx context.Context, d *conn.Dialer, p peer.ID, remote
 	// to end early.
 	go func() {
 		// rate limiting just in case. at most 10 addrs at once.
-		limiter := ratelimit.NewRateLimiter(process.Background(), 10)
+		limiter := ratelimit.NewRateLimiter(process.Background(), 8)
 		limiter.Go(func(worker process.Process) {
 			// permute addrs so we try different sets first each time.
 			for _, i := range rand.Perm(len(remoteAddrs)) {
-				select {
-				case <-foundConn: // if one of them succeeded already
-					break
-				case <-worker.Closing(): // our context was cancelled
-					break
-				default:
+				if conn.IsTcpMultiaddr(remoteAddrs[i]) {
+					// For addresses that require an FD to dial
+					select {
```

Comment:
Here, this fixes it all for you: 

```go
func (s *Swarm) addrDialRateLimit(a ma.Multiaddr) (chan<-  struct{}) {
  if isFDCostlyTransport(a) {
    return s.fdRateLimit
  }

  // do not rate limit it at all
  return make(chan struct{}, 1)
}

// and in here: 

for _, i := range rand.Perm(len(remoteAddrs)) {

  select {
  case <-foundConn: // if one of them succeeded already
    break
  case <-worker.Closing(): // our context was cancelled
    break
  default:
  }


  workerAddr := remoteAddrs[i] // shadow variable to avoid race

  // we have to do the waiting concurrently because there are addrs 
  // that SHOULD NOT be rate limited (utp), nor blocked by other 
  // rate limited addrs (tcp).

  // we need to call limiter.Go as required by goproc/limiter semantics.
  // note: limiter.Go is not limiter.LimitedGo.
  limiter.Go(func(p process.Process) {

    // returns whatever ratelimiting is acceptable for workerAddr.
    // may not rate limit at all.
    rl := s.addrDialRateLimit(workerAddr)
    rl<- struct{}{}

    limiter.LimitedGo(func(worker process.Process) {
      dialSingleAddr(workerAddr)
    })

    <-rl
  })

}
```